### PR TITLE
Add multi-hint functionality for buggy outputs, provided the student has changed their code.

### PR DESCRIPTION
**Pull request by eyurko**
Linted and three (count 'em) new tests added.

We should probably add a testing infrastructure, at some point ...

### CODE REVIEW

File: client/services/FeedbackGeneratorService.js [hunk 1]
```diff
@@ -55,6 +56,9 @@ tie.factory('FeedbackGeneratorService', [
 
     return {
       getFeedback: function(prompt, codeEvalResult) {
+        // TODO(eyurko): This is getting long enough that we should
```
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: I'd prefer to do this in a separate PR, and am happy to split it after this PR.

File: client/services/FeedbackGeneratorServiceSpec.js [hunk 2]
```diff
@@ -91,5 +98,159 @@ describe('FeedbackGeneratorService', function() {
         "3 seconds) we've set. Can you try to make it run ",
         "more efficiently?"].join(''));
     });
+
+    it(['should return the next hint in sequence for buggy outputs, ',
```
Comment (eyurko) [inline client/services/FeedbackGeneratorServiceSpec.js]: There's a fair bit of redundancy, here, and in subsequent tests -- would it be better to make this a global buggyOutputTestDict?

File: client/services/TranscriptService.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ tie.factory('TranscriptService', [
     var transcript = TranscriptObjectFactory.create();
 
     return {
+   	  getTranscript: function(){
```
Comment (seanlip) [inline client/services/TranscriptService.js]: Use spaces not tabs.

Comment (eyurko) [inline client/services/TranscriptService.js]: That's so weird -- it LOOKS fine on mine. Fixing.

File: client/domain/FeedbackObjectFactory.js [hunk 4]
```diff
@@ -22,6 +22,10 @@ tie.factory('FeedbackObjectFactory', [
     var Feedback = function(answerIsCorrect) {
       this._paragraphs = [];
       this._answerIsCorrect = answerIsCorrect;
+      // This records what message was displayed with this feedback.
+      // If no message (or no message with additional potential messages)
+      // was displayed, this will remain at -1.
```
Comment (seanlip) [inline client/domain/FeedbackObjectFactory.js]: Why not null? Some feedback isn't associated with hints anyway.

File: client/services/FeedbackGeneratorService.js (hunk 1 above)
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Mmm. How about pulling just the buggy function stuff out into a separate function now? Otherwise you have a giant block of code here -- and then the next PR will contain two giant blocks of code that are largely the same but hard to compare.

File: client/services/FeedbackGeneratorServiceSpec.js (hunk 2 above)
Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: Yes, I think so.

File: client/services/TranscriptService.js [hunk 5]
```diff
@@ -22,6 +22,9 @@ tie.factory('TranscriptService', [
     var transcript = TranscriptObjectFactory.create();
 
     return {
+      getTranscript: function(){
```
Comment (seanlip) [inline client/services/TranscriptService.js]: Add space before {

File: client/domain/FeedbackObjectFactory.js (hunk 4 above)
Comment (eyurko) [inline client/domain/FeedbackObjectFactory.js]: It lets me +1 it to 0, which is handy instead of having to have it as a separate case. The name suggests an int to me, anyways.

Comment (eyurko) [inline client/domain/FeedbackObjectFactory.js]: Although I don't actually do any work with it as an int, so I can change it to null if you'd prefer -- I usually just use -1 here because the name suggests an int, rather than an object to me.

File: client/services/TranscriptService.js (hunk 5 above)
Comment (eyurko) [inline client/services/TranscriptService.js]: Sure, done.

File: client/domain/FeedbackObjectFactory.js (hunk 4 above)
Comment (seanlip) [inline client/domain/FeedbackObjectFactory.js]: In general, null is like None in python, which is a fine value for ints/floats/strings/etc. properties that indicates the absence of a value. So it's totally fine to use it for fields of any given type. (Otherwise you wouldn't ever be able to use it...)

So I think null is fine; -1 seems a bit opaque to me and doesn't necessarily convey "none".

File: client/services/FeedbackGeneratorService.js [hunk 6]
```diff
@@ -53,8 +54,52 @@ tie.factory('FeedbackGeneratorService', [
...
+      if (lastSnapshot !== null) {
+        // This section makes sure to provide a new hint
+        // if the student gets stuck on the same bug by checking
+        // that they've submitted new code with the same error.
+        var previousFeedback = lastSnapshot.getFeedback();
+        var previousHintIndex = previousFeedback.getHintIndex();
+        if (previousHintIndex !== -1 && 
+          previousHintIndex < buggyMessages.length) {
+          var previousMessages = previousFeedback.getParagraphs();
+          // This could cause a problem if two different buggy outputs
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Can you make a note on the data schema doc that this is a requirement, so that we can include a check for it later in our schema validation code?

File: client/services/FeedbackGeneratorService.js [hunk 7]
```diff
@@ -53,8 +54,52 @@ tie.factory('FeedbackGeneratorService', [
...
+          // have the exact same hint, but that shouldn't be allowed.
+          if (previousMessages[0].getContent() == 
+            buggyMessages[previousHintIndex]) {
+            var previousCode = (
+              lastSnapshot.getCodeEvalResult().getCode());
+            if (previousCode === codeEvalResult.getCode()) {
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Can we just set the variable hintIndex here, and leave the construction of the feedback to outside the 'if' branch? That is: if the previous code is the same, set hintIndex to previousHintIndex, otherwise set it to 0 or previousHintIndex + 1. Construct the feedback object after that.

(Aim is to reduce some of the branching. It feels to me like a lot of this logic is just to figure out the correct hint index, and the actual construction of the feedback is orthogonal to that. Ditto for the "0" case later below.)

File: client/services/FeedbackGeneratorServiceSpec.js [hunk 8]
```diff
@@ -92,4 +99,149 @@ describe('FeedbackGeneratorService', function() {
...
+            "It looks like you're counting the number of parentheses, and if ",
+            "you have the same # of each kind, returning true. That's not ",
+            "quite correct. See if you can figure out why."
+          ].join('')
+        ]
+      };
+      var buggyOutputTest = BuggyOutputTestObjectFactory.create(
+        buggyOutputTestDict);
+      var codeEvalResult = CodeEvalResultObjectFactory.create(
+        'some code', 'same output', [], [true], [], false,
```
Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: Should the sixth arg be null, and the seventh arg be null, if there is no runtime error? Ditto below.

File: client/services/FeedbackGeneratorService.js (hunk 6 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: Done.

File: client/services/FeedbackGeneratorService.js (hunk 7 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: Sure. Fixed it.

File: client/services/FeedbackGeneratorServiceSpec.js (hunk 8 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorServiceSpec.js]: I think either way it doesn't trigger, but I'm fine modifying it.

Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: Yep, I agree it doesn't trigger (since we're not testing that part of the code in this test) but I think it's a good idea to make these calls reflect reality. Thanks!

File: client/services/FeedbackGeneratorService.js [hunk 9]
```diff
@@ -53,8 +54,46 @@ tie.factory('FeedbackGeneratorService', [
...
+      var lastSnapshot = (
+        TranscriptService.getTranscript().getPreviousSnapshot());
+      if (lastSnapshot !== null) {
+        // This section makes sure to provide a new hint
+        // if the student gets stuck on the same bug by checking
+        // that they've submitted new code with the same error.
+        var previousFeedback = lastSnapshot.getFeedback();
+        var previousHintIndex = previousFeedback.getHintIndex();
+        if (previousHintIndex !== null && 
+          previousHintIndex < buggyMessages.length) {
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Interesting -- what is supposed to happen if you're at the last hint of a set, and you change the code and still trigger the same error?

File: client/services/FeedbackGeneratorService.js [hunk 10]
```diff
@@ -53,8 +54,46 @@ tie.factory('FeedbackGeneratorService', [
...
+          // This could cause a problem if two different buggy outputs
+          // have the exact same hint, but that shouldn't be allowed.
+          if (previousMessages[0].getContent() == 
+            buggyMessages[previousHintIndex]) {
+            var previousCode = (
+              lastSnapshot.getCodeEvalResult().getCode());
+            if (previousCode === codeEvalResult.getCode()) {
+              hintIndex = previousHintIndex;
+            }
+            hintIndex = previousHintIndex + 1 % buggyMessages.length;
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Shouldn't there be an 'else' here? Otherwise you're just overwriting the previous value in all cases.

Also, not sure I understand the significance of the "%"?

File: client/services/FeedbackGeneratorService.js [hunk 11]
```diff
@@ -53,8 +54,46 @@ tie.factory('FeedbackGeneratorService', [
       }
     };
 
+    var _getBuggyOutputTestFeedback = function (
+      failingTest, codeEvalResult) {
+      var hintIndex = 0;
+      var buggyMessages = failingTest.getMessages();
+      var feedback = FeedbackObjectFactory.create(false);
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Minor: consider moving this below, near where you do all the feedback-related stuff.

File: client/services/FeedbackGeneratorServiceSpec.js [hunk 12]
```diff
@@ -92,4 +99,149 @@ describe('FeedbackGeneratorService', function() {
         "more efficiently?"].join(''));
     });
   });
+
+  describe('_getBuggyOutputTestFeedback', function() {
+    it(['should return the next hint in sequence for buggy outputs, ',
+        'provided the code has been changed'].join(''), function() {
+      var buggyOutputTestDict = {
```
Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: It might be worth declaring this more globally (in the 'describe', but outside the 'it') if you're going to reuse it a lot. Up to you, though.

File: client/services/FeedbackGeneratorServiceSpec.js [hunk 13]
```diff
@@ -92,4 +99,149 @@ describe('FeedbackGeneratorService', function() {
...
+        codeEvalResultWithNewError, unusedRuntimeErrorFeedback));
+
+      var feedback = FeedbackGeneratorService._getBuggyOutputTestFeedback(
+        buggyOutputTest, codeEvalResultWithSameBug).getParagraphs();
+
+      expect(paragraphs.length).toEqual(1);
+      expect(paragraphs[0].isTextParagraph()).toBe(true);
+      expect(paragraphs[0].getContent()).toBe(buggyOutputTestDict.messages[0]);
+    });
+  });
```
Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: Should we also have tests for the behavior near the end of the list of hints?

Comment (eyurko) [inline client/services/FeedbackGeneratorServiceSpec.js]: I added a test for making sure that the message repeats at the end of a list, but other than that it seems like it should behave the same throughout?

File: client/services/FeedbackGeneratorServiceSpec.js (hunk 12 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorServiceSpec.js]: It seems that way, so I will.

File: client/services/FeedbackGeneratorService.js (hunk 10 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: Ah, yes, sorry, forgot to change that when I rearchitected the method. The modulo was originally to wrap it back around to the start, but I think it's better for the student to see the same "most helpful" hint until they finish, so I've tweaked this.

File: client/services/FeedbackGeneratorService.js (hunk 9 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: At this point it'd loop around, hence the %, I believe. But it might be better to just stay on the last hint. I'll switch to that, but happy to change back if you prefer the other.

File: client/services/FeedbackGeneratorService.js (hunk 11 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: Sure, that works.

File: client/domain/FeedbackObjectFactory.js [hunk 14]
```diff
@@ -22,6 +22,10 @@ tie.factory('FeedbackObjectFactory', [
     var Feedback = function(answerIsCorrect) {
       this._paragraphs = [];
       this._answerIsCorrect = answerIsCorrect;
+      // This records what message was displayed with this feedback.
+      // If no message (or no message with additional potential messages)
```
Comment (eyurko) [inline client/domain/FeedbackObjectFactory.js]: Yup, I think this is from an old understanding of how this works.

File: client/services/FeedbackGeneratorService.js (hunk 9 above)
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: OK, then do you need the previousHintIndex < buggyMessages.length check? Would it not always be true?

File: client/services/FeedbackGeneratorService.js [hunk 15]
```diff
@@ -53,8 +54,48 @@ tie.factory('FeedbackGeneratorService', [
...
+          // have the exact same hint, but that shouldn't be allowed.
+          if (previousMessages[0].getContent() == 
+            buggyMessages[previousHintIndex]) {
+            var previousCode = (
+              lastSnapshot.getCodeEvalResult().getCode());
+            if (previousCode === codeEvalResult.getCode() ||
+                previousHintIndex == buggyMessages.length - 1) {
+              hintIndex = previousHintIndex;
+            } else {
+              hintIndex = previousHintIndex + 1 % buggyMessages.length;
```
Comment (seanlip) [inline client/services/FeedbackGeneratorService.js]: Why the `% buggyMessages.length`, if we are not looping around?

Also for future reference, % takes precedence over +.

File: client/services/FeedbackGeneratorServiceSpec.js (hunk 13 above)
Comment (seanlip) [inline client/services/FeedbackGeneratorServiceSpec.js]: Yep, sgtm. Thanks!

File: client/services/FeedbackGeneratorService.js (hunk 15 above)
Comment (eyurko) [inline client/services/FeedbackGeneratorService.js]: oop, that's fixed on my side. Must not have saved before committing. Sorry about that.

**Comment by seanlip** [comment]
Oh, @eyurko, sorry -- the previous PR merge caused this to have a merge conflict. Would you mind fixing it, and merging after that? Thanks!
